src/contract/deploy/impl/SourceImpl.ts

```diff
@@ -28,18 +28,31 @@ export interface SourceData {
 
 export class SourceImpl implements Source {
   private readonly logger = LoggerFactory.INST.create('Source');
-  constructor(private readonly arweave: Arweave) {}
+  private readonly wallet: Wallet;
+
+  constructor(private readonly warp: Warp) {
+    this.wallet = new Wallet(this.warp);
+  }
 
   async save(
     contractData: SourceData,
     env: WarpEnvironment,
-    signer: ArWallet | SigningFunction,
+    wallet: ArWallet | Signature,
     useBundler = false
   ): Promise<any> {
     this.logger.debug('Creating new contract source');
 
     const { src, wasmSrcCodeDir, wasmGlueCode } = contractData;
 
+    this.wallet.getSignature(wallet);
+    const signer = this.wallet.signature.signer;
+
+    if (this.wallet.signature.signatureType !== 'arweave' && !useBundler) {
```

Comment:
shouldn't this logic be extracted to the new Wallet class?

Reply:
done.